docs/design.md

```diff
@@ -155,88 +155,35 @@ Kata Container support has to be enabled explicitly via a [storage
-[Node controller server](#node-controller-server) running on a worker
-node that was registered with the driver.
-
-### Node Controller Server
-
-This gRPC server is started by the PMEM-CSI driver running in _Node_
-mode and implements the
-[ControllerPublishVolume](https://github.com/container-storage-interface/spec/blob/master/spec.md#controllerpublishvolume)
-and
-[ControllerUnpublishVolume](https://github.com/container-storage-interface/spec/blob/master/spec.md#controllerunpublishvolume)
-methods of the [Controller
-service](https://github.com/container-storage-interface/spec/blob/master/spec.md#controller-service-rpc)
-interface defined by the CSI specification. It serves the
-ControllerPublishVolume() and ControllerUnpublish() requests coming
-from the [Master controller server](#master-controller-server) and
-creates/deletes persistent memory devices.
-
-### Node Server
-
-This gRPC server is started by the driver running in _Node_ mode and
-implements the [Node
-service](https://github.com/container-storage-interface/spec/blob/master/spec.md#node-service-rpc)
-interface defined in the CSI specification. It serves the
-NodeStageVolume(), NodeUnstageVolume(), NodePublishVolume(), and
-NodeUnpublishVolume() requests coming from the Container Orchestrator
-(CO).
+## Dynamic provisioning of local volumes
+
+Traditionally, Kubernetes expects that a driver deployment has a
+central component, usually implemented with the `external-provisioner`
+and a custom CSI driver component which implements volume creation.
+That central component is hard to implement for a CSI driver that
+creates volumes locally on a node.
+
+PMEM-CSI solves this problem by deploying `external-provisioner`
+alongside each node driver and enabling ["distributed
+provisioning"](https://github.com/kubernetes-csi/external-provisioner/issues/487):
+- For volumes with storage classes that use late binding (aka "wait
+  for first consumer"), a volume is tentatively assigned to a node
+  before creating it, in which case the `external-provisioner` running
+  on that node can tell that it is responsible for provisioning.
+- For volumes with storage classes that use immediate binding, the
+  different `external-provisioner` instances compete with each for
+  ownership of the volume by setting the "selected node"
+  annotation. Delays are used to avoid the thundering herd problem.
+  Once a node has been selected, provisioning continues as with late
+  binding. This is less efficient and therefore "late binding" is the
+  recommended binding mode.
+
+PMEM-CSI also has a central component which implements the [scheduler
+extender](#scheduler-extender) webhook. That component needs to know
+on which nodes the PMEM-CSI driver is running and how much capacity is
+available there. This information is retrieved by dynamically
+discovering PMEM-CSI pods and connecting to their [metrics
+endpoint](/docs/install.md#metrics-support).
```

Comment:
This makes 'enabling the metrics support' a hard dependency on 'scheduler extension'. The code should implement this by always enabling node metrics when scheduler extension is enabled.

Reply:
Does the node part know whether scheduler extensions are enabled? I don't think so, because those parameters are passed to the controller part.

Reply:
Or did you mean some other code than the main.go and pmem-csi-driver.go?

Reply:
> Does the node part know whether scheduler extensions are enabled? I don't think so, because those parameters are passed to the controller part.

Yes, your are true, it is not possible to enforce this in the code. Would be good to add a note about "enabling metrics while using scheduler extension" in the documentation.

Reply:
It's documented that the scheduler extension code uses metrics. I think that is enough. We don't describe how to disable metrics support (it's always on in YAML and operator deployments), so the risk that someone accidentally deploys incorrectly is low.